go/test/endtoend/vreplication/cluster.go

```diff
@@ -106,6 +109,128 @@ func setTempVtDataRoot() string {
+// ./bin, ./sbin, and ./libexec subdirectories of VT_MYSQL_ROOT.
+func setVtMySQLRoot(mysqlRoot string) error {
+	if _, err := os.Stat(mysqlRoot); os.IsNotExist(err) {
+		os.Mkdir(mysqlRoot, 0700)
+	}
+	err := os.Setenv("VT_MYSQL_ROOT", mysqlRoot)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("VT_MYSQL_ROOT is %s\n", mysqlRoot)
+	return nil
+}
+
+// setDBFlavor sets the MYSQL_FLAVOR OS env var.
+// You should call this after calling setVtMySQLRoot() to ensure that the
+// correct flavor is used by mysqlctl based on the current mysqld version
+// in the path. If you don't do this then mysqlctl will use the incorrect
+// config/mycnf/<flavor>.cnf file and mysqld may fail to start.
+func setDBFlavor() error {
+	versionStr, err := mysqlctl.GetVersionString()
+	if err != nil {
+		return err
+	}
+	f, v, err := mysqlctl.ParseVersionString(versionStr)
+	if err != nil {
+		return err
+	}
+	flavor := fmt.Sprintf("%s%d%d", f, v.Major, v.Minor)
+	err = os.Setenv("MYSQL_FLAVOR", string(flavor))
+	if err != nil {
+		return err
+	}
+	fmt.Printf("MYSQL_FLAVOR is %s\n", string(flavor))
+	return nil
+}
+
+func unsetVtMySQLRoot() {
+	_ = os.Unsetenv("VT_MYSQL_ROOT")
+}
+
+func unsetDBFlavor() {
+	_ = os.Unsetenv("MYSQL_FLAVOR")
+}
+
+// getDBTypeVersionInUse checks the major DB version of the mysqld binary
+// that mysqlctl would currently use, e.g. 5.7 or 8.0 (in semantic versioning
+// this would be major.minor but in MySQL it's effectively the major version).
+func getDBTypeVersionInUse() (string, error) {
+	var dbTypeMajorVersion string
+	versionStr, err := mysqlctl.GetVersionString()
+	if err != nil {
+		return dbTypeMajorVersion, err
+	}
+	flavor, version, err := mysqlctl.ParseVersionString(versionStr)
+	if err != nil {
+		return dbTypeMajorVersion, err
+	}
+	majorVersion := fmt.Sprintf("%d.%d", version.Major, version.Minor)
+	if flavor == mysqlctl.FlavorMySQL || flavor == mysqlctl.FlavorPercona {
+		dbTypeMajorVersion = fmt.Sprintf("mysql-%s", majorVersion)
+	} else {
+		dbTypeMajorVersion = fmt.Sprintf("%s-%s", strings.ToLower(string(flavor)), majorVersion)
+	}
+	return dbTypeMajorVersion, nil
+}
+
+// downloadDBTypeVersion downloads a recent major version release build for the specified
+// DB type.
+// If the file already exists, it will not download it again.
+// The artifact will be downloaded and extracted in the specified path. So e.g. if you
+// pass /tmp as the path and 5.7 as the majorVersion, mysqld will be installed in:
+// /tmp/mysql-5.7/bin/mysqld
+// You should then call setVtMySQLRoot() and setDBFlavor() to ensure that this new
+// binary is used by mysqlctl along with the correct flavor specific config file.
+func downloadDBTypeVersion(dbType string, majorVersion string, path string) error {
+	client := http.Client{
+		Timeout: 10 * time.Minute,
+	}
+	var url, file, versionFile string
+	dbType = strings.ToLower(dbType)
+
+	if dbType == "mysql" && majorVersion == "5.7" {
+		versionFile = "mysql-5.7.37-linux-glibc2.12-x86_64.tar.gz"
```

Comment:
This is gonna crash in OSX. You should skip the test if not being ran on Linux to avoid weird behaviors.

Reply:
Yeah, on x86_64 linux specifically. I'll add a check and log.Fatal for this. Thanks for the reminder! I was so focused on the CI ...